source/_posts/2019-01-28-building-a-decentralized-reddit-with-embark-part-2.md

```diff
@@ -0,0 +1,194 @@
+title: Building a decentralized Reddit with Embark - Part 2
+summary: "This is the second part of the three part tutorial about building a decentralized Reddit with Embark. In this part, we'll be focussing on testing our Smart Contract using EmbarkJS."
+categories:
+  - tutorials
+---
+
+In [the first part of this tutorial](/news/2019/01/28/building-a-decentralized-reddit-with-embark-part-1/) we've implemented a `DReddit` Smart Contract that comes with methods to create and vote on topic posts. In this part we'll continue right where we've left off and take a closer look at how we can test our Smart Contract using Embark.
+
+And off we go!
+
+## Writing a first test
+
+We've got plenty functionality to cover in our tests, but let's start with a very simple one just to get a bit more familiar with how to write tests and also to ensure things are working as intended. First we create a test file `DReddit_spec.js` inside `test` and add a `contract()` block that looks something like this:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+
+});</code></pre>
+
+Inside this code block we'll be putting dedicated test cases. The `contract()` function can be considered a "grouping" functionality to group tests, if you will. It comes with a couple of characteristics specific to Embark, but generally one can say that `contract()` is not only the equivalent to, but an extension of Mocha's [describe()]() function.
+
+To check whether our test setup is working, we add a simple test that passes:
+
+<pre class="javascript"><code>contract('DReddit', () => {
+
+  it ('should work', () => {
+    assert.ok(true);
+  });
+});</code></pre>
+
+Running this using Embark's `test` command should result in an output similar to this:
+
+<pre class="shell"><code>❯ embark test
+
+
+Compiling contracts
+  DReddit
+    ✓ should work (0ms) - [0 gas]
+
+
+  1 passing (5s) - [Total: 2210775 gas]
+
+ > All tests passed
+</code></pre>
+
+This works great, let's go ahead and test some actual functionality!
+
+## Testing the creation of post
+
+Let's test the core functionality of our application - the creation of posts. For that we need to do a couple of things: We need to somehow get an instance of our `DReddit` Smart Contract in JavaScript, so we can call methods on it to test if they work, and we also need to configure out testing environment so that the right Smart Contract instances are created.
+
+### Requiring Smart Contract instances
+
+When running tests, Embark adds a couple of custom functions and objects to the global scope, which are useful or even necessary. One of those functions is a custom `require()` that lets us import Smart Contract instances from an Embark specific path.
```

Comment:
which are ~~useful or even~~ necessary
We could probably explain why it is necessary or link to the doc that explains it

Reply:
Hm.. yes, but it's actually only necessary because we're providing those custom paths to import from. But theoretically, we could have `embark test` output compiled artifacts on the disc as well and have users import from there. So the necessity kind of comes with the fact that we introduce this thing in the first place.

Is there any other reason we introduce a custom `require()`?